packages/babel-parser/src/parser/statement.js

```diff
@@ -1014,6 +1014,15 @@ export default class StatementParser extends ExpressionParser {
       this.unexpected(null, "let is disallowed as a lexically bound name");
     }
     decl.id = this.parseBindingAtom();
+    if (kind === "const" || kind === "let") {
+      const invalid = this.checkVarIdHasLet(decl.id);
+      if (invalid) {
+        this.raise(
+          invalid.start,
+          "'let' is not allowed to be used as a name in 'let' or 'const' declarations.",
+        );
+      }
+    }
     this.checkLVal(
```

Comment:
If it works, I would prefer to move the check to `checkLVal`, to avoid looping over the destructuring tree twice.
We can check if `bindingType === BIND_LEXICAL` when checking identifiers, or pass an additional `disallowLetBinding` parameter to that function.